cmd/registration-tracker/main.go

```diff
@@ -0,0 +1,377 @@
+	}
+
+	regstats := regStats{
+		logger:          logger,
+		m:               &sync.Mutex{},
+		v4Registrations: make(map[uint]*transportStats),
+		v6Registrations: make(map[uint]*transportStats),
+		currRegIds:      make(map[string]int),
+		prevRegIds:      make(map[string]int),
+		GeoIP:           geoipDB,
+	}
+
+	ctx, _ := context.WithCancel(context.Background())
+	regChan := make(chan interface{}, 10000)
+	zmqIngester, err := cj.NewZMQIngest(zmqAddress, regChan, privkey, conf.ZMQConfig)
+	if err != nil {
+		logger.Fatal("error creating ZMQ Ingest: %w", err)
+	}
+
+	// Receive registration updates from ZMQ Proxy as subscriber
+	go zmqIngester.RunZMQ(ctx)
+
+	regManager := cj.NewRegistrationManager(&cj.RegConfig{})
+
+	// Add supported transport options for registration validation
+	for transportType, transport := range enabledTransports {
+		err = regManager.AddTransport(transportType, transport)
+		if err != nil {
+			logger.Fatalf("failed to add transport: %v", err)
+		}
+	}
+
+	ticker := time.NewTicker(30 * time.Second)
+	go func() {
+		for range ticker.C {
+			regstats.PrintAndReset()
+		}
+	}()
+
+	msgChan := make(chan interface{}, 10000)
+
+	// Read from regChan to which the zmqIngester writes
+	for data := range regChan {
+
+		select {
+		case <- ctx.Done():
+			logger.Fatal("closing all ingest threads")
+		case msgChan <- data:
+		default:
+			logger.Tracef("dropping registration")
+		}
+		msg, ok := (<-msgChan).([]byte)
+		if !ok {
+			logger.Tracef("failed to convert registration into proper type")
+		}
+
+		parsed := &pb.C2SWrapper{}
+		err = proto.Unmarshal(msg, parsed)
+		if err != nil {
+			logger.Errorf("failed to unmarshall ClientToStation: %v", err)
+		}
+
+		if parsed.GetRegistrationAddress() == nil {
+			parsed.RegistrationAddress = make([]byte, 16)
+		}
+		if parsed.GetDecoyAddress() == nil {
+			parsed.DecoyAddress = make([]byte, 16)
+		}
+
+		var sourceAddr = net.IP(parsed.GetRegistrationAddress())
+
+		var reg *cj.DecoyRegistration
+
+		// gotapdance cli support v6 by default, so Getv6Support() will typically be true
+		// Also, clients IP version support can either be (v6 and v4) or v4 alone
+		if parsed.GetRegistrationPayload().GetV6Support() {
+			// registration indicates support for v6 
+			reg, err = regManager.NewRegistrationC2SWrapper(parsed, true)
+			if err != nil {
+				handleC2SError(logger, "v6-supported", err, reg)
+				continue
+			}
+		} else if parsed.GetRegistrationPayload().GetV4Support() && sourceAddr.To4() != nil {
+			// registration indicates NO support for v6 
+			reg, err = regManager.NewRegistrationC2SWrapper(parsed, false)
+                        if err != nil {
+                                handleC2SError(logger, "strictly-v4-supported", err, reg)
+                                continue
+                        }
+		} else {
+			logger.Errorf("reached an edge case where the client neither supports v4 or v6")
+			continue
+		}
+
+		go func(ip net.IP, registration *cj.DecoyRegistration) {
```

Comment:
You may want to turn this into a worker model, rather than launching a new thread per registration. New threads have the downside that you may launch too many (if a storm of new registrations comes in), and there's likely some upper-bound of useful threads doing this.

If this tool never runs on a production server it's probably okay, but if it does we would want to bound the number of concurrent threads launched by this tool to prevent us from consuming too much CPU.